drivers/i3c/i3c_shell.c

```diff
@@ -1194,6 +1262,142 @@ static int cmd_i3c_ccc_getcaps(const struct shell *shell_ctx, size_t argc, char
 	return ret;
 }
 
+/* i3c ccc getvendor <device> <target> <id> [<defining byte>] */
+static int cmd_i3c_ccc_getvendor(const struct shell *shell_ctx, size_t argc, char **argv)
+{
+	const struct device *dev, *tdev;
+	struct i3c_device_desc *desc;
+	uint8_t buf[MAX_I3C_BYTES] = {0};
+	uint8_t defbyte;
+	size_t num_xfer;
+	uint8_t id;
+	int ret;
+
+	dev = device_get_binding(argv[ARGV_DEV]);
+	if (!dev) {
+		shell_error(shell_ctx, "I3C: Device driver %s not found.", argv[ARGV_DEV]);
+		return -ENODEV;
+	}
+	tdev = device_get_binding(argv[ARGV_TDEV]);
+	if (!tdev) {
+		shell_error(shell_ctx, "I3C: Device driver %s not found.", argv[ARGV_TDEV]);
+		return -ENODEV;
+	}
+	desc = get_i3c_attached_desc_from_dev_name(dev, tdev->name);
+	if (!desc) {
+		shell_error(shell_ctx, "I3C: Device %s not attached to bus.", tdev->name);
+		return -ENODEV;
+	}
+
+	id = strtol(argv[3], NULL, 16);
+	if (id > 0x1E) {
+		shell_error(shell_ctx, "I3C: CCC VENDOR ID %d is out of range.", id);
+		return -EINVAL;
+	}
+
+	if (argc > 3) {
+		defbyte = strtol(argv[3], NULL, 16);
+		ret = i3c_ccc_do_getvendor_defbyte(desc, id, defbyte, buf, MAX_I3C_BYTES,
+						   &num_xfer);
+	} else {
+		ret = i3c_ccc_do_getvendor(desc, id, buf, MAX_I3C_BYTES, &num_xfer);
+	}
+
+	if (ret < 0) {
+		shell_error(shell_ctx, "I3C: unable to send CCC VENDOR.");
+		return ret;
+	}
+
+	shell_hexdump(shell_ctx, buf, num_xfer);
+
+	return ret;
+}
+
+/* i3c ccc setvendor <device> <target> <id> [<bytes>] */
+static int cmd_i3c_ccc_setvendor(const struct shell *shell_ctx, size_t argc, char **argv)
+{
+	const struct device *dev, *tdev;
+	struct i3c_device_desc *desc;
+	struct i3c_driver_data *data;
+	uint8_t buf[MAX_I3C_BYTES] = {0};
+	uint8_t data_length;
+	uint8_t id;
+	int ret;
+	int i;
+
+	dev = device_get_binding(argv[ARGV_DEV]);
+	if (!dev) {
+		shell_error(shell_ctx, "I3C: Device driver %s not found.", argv[ARGV_DEV]);
+		return -ENODEV;
+	}
+	tdev = device_get_binding(argv[ARGV_TDEV]);
+	if (!tdev) {
+		shell_error(shell_ctx, "I3C: Device driver %s not found.", argv[ARGV_TDEV]);
+		return -ENODEV;
+	}
+	desc = get_i3c_attached_desc_from_dev_name(dev, tdev->name);
+	if (!desc) {
+		shell_error(shell_ctx, "I3C: Device %s not attached to bus.", tdev->name);
+		return -ENODEV;
+	}
+	data = (struct i3c_driver_data *)dev->data;
+
+	id = strtol(argv[3], NULL, 16);
+	if (id > 0x1E) {
+		shell_error(shell_ctx, "I3C: CCC VENDOR ID %d is out of range.", id);
+		return -EINVAL;
+	}
```

Comment:
why is shell double-checking here? I3C code will already fail.

Reply:
it's to give a specific error print back to the user

Reply:
I don't think this is a good error handling practice, `-EINVAL` can be documented as a value returned for that condition. It's true that it is difficult to be precise just with errno, but that's what we have... Logging in the driver can also help (if activated, user will get more precise info)

Reply:
removed